Claudine/addrandom

claudinepeyrat06 opened the pull request:
This is an attempt to add some randomness to some of the stitches. It is also my first trial at modifying the code...

I'm fairly happy with what i did with the satin columns, much  less sure about the  fills....


![IMG_2227](https://user-images.githubusercontent.com/88194877/195581323-e065c0ec-29fe-4551-bcce-4f20568b78bf.jpg).

Commit:
satin colomn

add randomness
allow pull compensation to be a percentage
correct reverse in the case of satin column with no rung.

Commit:
fill stitch randomness

allow some randomness on fill stitches : autofill, contour fill and guided fill

Commit:
add zigzag spacing  randomness

Commit:
bug  corrections

bug correction

Commit:
Update satin_column.py

claudinepeyrat06 commented:
Yes , as you can better see here
<img width="339" alt="image" src="https://user-images.githubusercontent.com/88194877/195630023-cea51e38-52df-4838-90cf-d768ed990558.png">

I added parameters  to add randomness to :
- zigzag spacing
- the position of the split when  a max stitch length is applied
- for each zig (or zag) how far from the rail can the end point be (for each rail, decide what max percentage of increase or decrease you allow).

Maybe it is too many parameters....

Augusa commented:
May be it is difficult to find the good params. I have a wonderfull file from a panda and the fur is made in this way.

george-steel commented on lib/elements/satin_column.py:
```diff
@@ -132,6 +174,32 @@ def zigzag_spacing(self):
+
+
+    @property
+    @param(
+        'pull_compensation_unit',
+        _('Pull compensation_unit'),
+        tooltip=_('can be either mm or percentage'),
+        type='dropdown',
```
It would be cleaner to add a `pull_compensation_percent` field. Pull compensation in mm and in percent compensate for different properties of the fabric and it might be useful to have both.

claudinepeyrat06 replied:
I'll change that.

george-steel commented on lib/elements/satin_column.py:
```diff
@@ -747,8 +834,32 @@ def add_pair(pos0, pos1):
+                    if self.random_first_rail_factor_out:
+                        increase=self.random_first_rail_factor_out
+                    if self.random_first_rail_factor_in or self.random_first_rail_factor_out:
+                        decalage0= random.uniform(-decrease/100,increase/100)
```
I don't like the fact that this will produce different results between runs. Perhaps we should use a seeded (perhaps with a hash of the element ID) PRNG instead.

claudinepeyrat06 replied:
I had been wondering that.... on the other hand it might be nice to be able to try again if  one is not happy with the first result.

kaalleen replied:
I agree with both of you. But once you are happy with the outcome, you will want you could keep it.
Maybe we can save the hash as an attribute to the element and add a button to recalculate?

george-steel replied:
Now I'm thinking that we seed the PRNG from the element ID, but have a param to override the seed. That way we could try different seeds if we need to and then keep a result we're happy with.

As for PRNGs, I am going to try ChaCha20 as it is random-access with a 128-bit position field, enough to simply concatenate loop indexes together so that slight changes in the shape's geometry won't completely reorder random calls.

george-steel commented:
@claudinepeyrat06  This looks like a great start. I've been meaning to add some randomization to satins and fills myself. Do you mind if I copy this into my own branch and play around with some possible improvements?

Commit:
pull compensation percentage

allow pull compensation percentage and fixed pull compensation

Commit:
style correction

style correction

Commit:
fix random split factor

Commit:
fix e_stitch bug

Commit:
add pull_compensation_rails parameter

decide wether the compensation should be applied both side or only on a chosen side

Commit:
Update satin_column.py

lexelby commented:
This is cool!

I have an idea for how to make this even better.  I would like it if the user could press a button to re-randomize the random parts.  The code could use a seeded random number generator so that the random parts are always generated the same until the user presses the randomize button.  This is something I was already thinking of doing for a feature I've been tinkering with.

bianchifan commented:
I'm not happy with the results when using satin stitches.
The width of the zigzag seems to be defined by the rungs.
In former versions (2.1) it was defined by the lines.

An extract from my svg in inkscape
![Sports_Logo_partly](https://user-images.githubusercontent.com/17491743/196625696-5ee63665-a360-465e-9446-412af9be311c.jpg)

How it looks with v 2.1
![Sports_2 1](https://user-images.githubusercontent.com/17491743/196625841-a694bc34-ebd1-423b-bbbb-913da56fc67f.jpg)

The result with add_random
![Sports_add_random](https://user-images.githubusercontent.com/17491743/196625947-91e85ba2-47e8-4e66-9612-814ec7c54a00.jpg)

Commit:
correct pull_compensation

correct pull_compensation

Commit:
clean import

claudinepeyrat06 commented:
Thanks for your feed back. 
You were right, there was something wrong with the  pull compensation. Hopefully it's fixed.

Commit:
correct yet another bug

correct pb with split satin underlay

Commit:
Update satin_column.py

Commit:
Update satin_column.py

bianchifan commented:
Looks fine now, THX!

Commit:
add seed information on satin column

for each satin column store the seed value of the pseudo random generator
add a button to  change the seed

Commit:
add seed to fill stitch

add seed to fill stitch

george-steel commented on lib/elements/satin_column.py:
```diff
@@ -1052,6 +1056,16 @@ def to_stitch_groups(self, last_patch=None):
 
+        # If no seed is defined, compute one randomly using time to seed,  otherwise, use stored seed
+
+        if self.use_seed == 0:
+            random.seed()
+            x = random.randint(1, 10000)
+            random.seed(x)
+            self.set_param("use_seed", x)
+        else:
+            random.seed(self.use_seed)
```
I would rather not rely on a hidden global like this. I'm working on a PR that uses the stateless interface from `hashlib` instead of `random` and maintains a counter in a local variable.

claudinepeyrat06 replied:
looking forward to see your work . I'm sure I will learn a lot from it... Thanks for that opportunity.

lexelby replied:
I'm looking forward to seeing what both of you come up with for this.  I had envisioned it as an actual user-settable parameter, visible in the Params window, maybe with a randomize button much like how Inkscape's live path effects work.  The UI may be a little tricky though.

george-steel replied:
The hashlib version will help with that since it will allow seeds to be arbitrary strings.

Commit:
revisiting fill stitch

hopefully a better approach to fill stitch.
feathering is first dealt with so as to modify the shape.

george-steel commented:
@claudinepeyrat06, as there are a bunch of subtle issues regarding the interaction of randomization and the stagger system in fill stitches, would you be open to breaking this PR into a couple of smaller ones, to do the satin stitch and the fill stitches separately? I'm happy to help you squash and split it if you like.

claudinepeyrat06 commented:
@george-steel  I can try do to that, but probably  only mid November when i will be back from a one week vacation.
In this new version of fill stitch randomization i think that i have better dealt with the stagger system.

claudinepeyrat06 commented:
![frandomchalo](https://user-images.githubusercontent.com/88194877/200179769-cc04cf7b-bd30-466f-9b2e-abae58dd1e6c.jpg)

to show some randomized fll

george-steel commented:
> @george-steel I can try do to that, but probably only mid November when i will be back from a one week vacation. In this new version of fill stitch randomization i think that i have better dealt with the stagger system.

@claudinepeyrat06 I can squash, split, and push to a new branch (perhaps `claudine/addrandom-squashed`). I'd like to work on my own improvements one stitch type at a time so basing them off a squashed and split version would be a lot easier. Is that OK with you?

claudinepeyrat06 commented:
@george-steel , Sure, please do it.

george-steel commented:
@claudinepeyrat06 I've squashed, split and pushed to [`claudine/addrandom-squashed`](https://github.com/inkstitch/inkstitch/tree/claudine/addrandom-squashed] and I will base my changes off the commits in that branch. If you want to update this PR to it, checkout that branch and force-push it here.

george-steel commented:
@claudinepeyrat06 An improved version of the pull compensation changes is up at #1897 which branches off of [claudine/addrandom-squashed](https://github.com/inkstitch/inkstitch/tree/claudine/addrandom-squashed). I could use your feedback on it.

Augusa commented:
I was very happy to use this new feature. I embrodered three motifs.
The first one 
![20221103_201010](https://user-images.githubusercontent.com/54963659/201526879-02c409db-2b05-4f7e-81a8-5cd9d6ec9f33.jpg)
I use randomised comlumns.
Then the second 
![20221108_183942](https://user-images.githubusercontent.com/54963659/201526959-2965a16d-8ecb-4dc7-bd4e-389b91e63e1b.jpg)
I used another fill on the leg.
On the third one I add randomised fills. As I didnot add  compensation I have some holes on the rand of my objects.
![20221109_184042](https://user-images.githubusercontent.com/54963659/201527183-3b4111ba-765e-428c-a2e3-161f8319c769.jpg)
Finally I like very much this new feature.

lexelby commented:
That first squirrel is AMAZING, I'm so happy Ink/Stitch can do that!

george-steel commented:
@Augusa would you mind sharing the svgs for the squirrel? They would make good tests.

Augusa commented:
It is the last one. Some of my files are not working with the last version and I have forgotten which files and witch version! 
![ecureuil-2022reduit-5bis](https://user-images.githubusercontent.com/54963659/201543386-ce863b19-215b-4cdf-b3b4-562b04f7d2e4.svg)

bianchifan commented:
Excellent work!
You were faster than me, I started converting some squirrel pics but switched to a "1day-butterfly" for diving into the ripple stitch.
Now I'm waiting for MAdeira Avalon to stitch it out..

@Augusa 
I took your "Ripple Bird" last week and reworked it for my EU5 (13x20)
Am I allowed to post it on a german sewing forum?

@claudinepeyrat06
Same question, I did an Inkscape rework from your bee pics, am I allowed to post them if they do stitch out properly?

claudinepeyrat06 commented:
@bianchifan of course you may

claudinepeyrat06 committed:
add pull_compensation for fill

claudinepeyrat06 committed:
pull_compensation on fill stitch

claudinepeyrat06 commented:
I have added fill compensation for fill stitch (two values, one for "left" side of rows, one for the other side),
 and also instead of a single parameter for the feathering effect,4 parameters to control how much  one wants inside, outside, and distinguish between the left side and right side of rows.
This should allow some user test to check what parameters are really useful for the end user.....

claudinepeyrat06 committed:
update guided_fill

claudinepeyrat06 committed:
Merge branch 'main' into claudine/addrandom

bianchifan commented:
@claudinepeyrat06 THX!

@Augusa
I mean your [paon from the InkStitch Tuts](https://inkstitch.org/tutorials/ripple-bird/)

claudinepeyrat06 commented:
@bianchifan , that bird is mine, so feel free to use it....

claudinepeyrat06 committed:
update fill_stitch.py

add a final call to buffer to clean up the shape when pull compensation is applied

george-steel commented:
> Yes , as you can better see here <img alt="image" width="339" src="https://user-images.githubusercontent.com/88194877/195630023-cea51e38-52df-4838-90cf-d768ed990558.png">

@claudinepeyrat06 would you mind posting the SVG for this so I can test it?

bianchifan commented:
> would you mind posting the SVG for this so I can test it?

You can build it by your own, it easy ;)
![Bee](https://user-images.githubusercontent.com/17491743/206922961-c5abd4b0-50f4-4ede-80dc-c8428bc61346.svg)

kaalleen commented:
Closing in favor of #1918 which has been already merged to main.
